Adapt EC2 Puppet handling if address_of an interface is None? (mypy)

neiljp opened the pull request:
Currently, address_of may return None, and is annotated as such.

In the first commit I've amended guess_gateway such that it may also return None, if the address is None. (It might be simpler if guess_gateway had an address (rather than device) as parameter, but that's a side issue) 

Both functions are subsequently used explicitly in a block where the address is explicitly None (line 139 in the patched file). Since I'm not entirely familiar with what the code is trying to solve, I'm unsure if the "Horrible hack" section should be outside of that block and so might work in some cases, or if the dhcpd line (if it succeeds) should assign an address? The second commit assumes the latter and pre-assigns the result of these functions and asserts that they are not None, which is the simplest resolution I can imagine if dhcpd successfully gives an IP address.

Thoughts?

(mypy also complains regarding the to_configure.remove line below the block, since there is no guarantee at that position that the passed-in parameter from address_of is not None, so the above could be adjusted to resolve this too)

smarx commented:
*Automated message from Dropbox CLA bot*

@neiljp, it looks like you've already signed the Dropbox CLA. Thanks!

timabbott commented on puppet/zulip_ops/files/zulip-ec2-configure-interfaces:
```diff
@@ -65,11 +65,15 @@ def address_of(device_id):
+    address = address_of(device_id)
+    if address is None:
+        return None
+    else:
```
Remove this `else` after the `return`; the code will read lot cleaner that way.

timabbott commented:
@neiljp these changes reasonable; posted one comment.  I'm a bit wary of modifying this script because it's not heavily used and kinda a pain to test, so I think we should restrict changes here to getting mypy passing (as opposed to generally refactoring it to be sane), since I think that refactoring would have the unfortunate setup of being high-risk and low-value.

neiljp commented:
Thanks @timabbott. I can see how infrequent use can make this difficult to test, though the changes should at least be python-only (ie. not affecting how commands are passed, etc)

I've amended the first commit as per your comment and adjusted the commit titles (if EC2 is specific enough).

I also added a third commit which resolves the last mypy (strict-optional) response to this file for me, which also looks straightforward (`to_configure.remove(None)` would raise an exception, which isn't handled?).

Anyhow, back to you :)

timabbott commented:
Looks great, merged, thanks @neiljp!

How many more errors do we have in `mypy --strict-optional`?  I'd love to get that list down to 0 so we can start enforcing it by default.
